tunnel/intercept/ebpf/ebpf_linux.go

```diff
@@ -0,0 +1,520 @@
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package ebpf
+
+import (
+	"context"
+	"fmt"
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/edge/tunnel"
+	"github.com/openziti/edge/tunnel/dns"
+	"github.com/openziti/edge/tunnel/entities"
+	"github.com/openziti/edge/tunnel/intercept"
+	"github.com/openziti/edge/tunnel/router"
+	"github.com/openziti/edge/tunnel/udp_vconn"
+	"github.com/openziti/foundation/v2/info"
+	"github.com/openziti/foundation/v2/mempool"
+	"github.com/openziti/foundation/v2/stringz"
+	"github.com/openziti/sdk-golang/ziti/edge/impl"
+	cmap "github.com/orcaman/concurrent-map/v2"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+	"net"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+// https://github.com/torvalds/linux/blob/master/Documentation/networking/tproxy.txt
+
+// Configure listening sockets with options that must be set before the socket is bound to an address (IP_TRANSPARENT).
+var listenConfig = net.ListenConfig{
+	Control: func(network, address string, c syscall.RawConn) error {
+		var sockOptErr error
+		controlErr := c.Control(func(sockFd uintptr) {
+			// - https://www.kernel.org/doc/Documentation/networking/tproxy.txt
+			if err := unix.SetsockoptInt(int(sockFd), unix.IPPROTO_IP, unix.IP_TRANSPARENT, 1); err != nil {
+				sockOptErr = fmt.Errorf("error setting IP_TRANSPARENT socket option: %v", err)
+				return
+			}
+			if err := unix.SetsockoptInt(int(sockFd), unix.SOL_SOCKET, unix.SO_REUSEADDR, 1); err != nil {
+				sockOptErr = fmt.Errorf("error setting SO_REUSEADDR socket option: %v", err)
+				return
+			}
+
+			if err := unix.SetsockoptInt(int(sockFd), syscall.SOL_IP, unix.IP_RECVORIGDSTADDR, 1); err != nil {
+				sockOptErr = fmt.Errorf("error setting SO_REUSEADDR socket option: %v", err)
+				return
+			}
+		})
+		if controlErr != nil {
+			return fmt.Errorf("error invoking listener socket control function: %v", controlErr)
+		}
+		return sockOptErr
+	},
+}
+
+func New() (intercept.Interceptor, error) {
+	return &interceptor{
+		serviceProxies: cmap.New[*eBpf](),
+	}, nil
+}
+
+type alwaysRemoveAddressTracker struct{}
+
+func (a alwaysRemoveAddressTracker) AddAddress(string) {}
+
+func (a alwaysRemoveAddressTracker) RemoveAddress(string) bool {
+	return true
+}
+
+type interceptor struct {
+	serviceProxies cmap.ConcurrentMap[*eBpf]
+}
+
+func (self *interceptor) Stop() {
+	self.serviceProxies.IterCb(func(key string, proxy *eBpf) {
+		proxy.Stop(alwaysRemoveAddressTracker{})
+	})
+	self.serviceProxies.Clear()
+}
+
+func (self *interceptor) Intercept(service *entities.Service, resolver dns.Resolver, tracker intercept.AddressTracker) error {
```

Comment:
Might want to add a check here to make sure that any command line utility you need is present, executable, has the right version, etc, and panic if there's a problem.

Reply:
Makes sense thanks for the feedback

Reply:
Hi Paul I added error checking for external map_update binary and associated ebpf maps in "func New()" since that is more consistent with where we check if iptables is available in tproxy mode.